File: companion/Protocol.cs
using System.Security.Cryptography;
using System.Text;
using Microsoft.AspNetCore.Http;

namespace JonaHomelab.Companion;

public static class Protocol
{
    public const string Version = "1";
    public const string RequestSignatureHeader = "X-Jona-Signature";
    public const string TimestampHeader = "X-Jona-Timestamp";
    public const string NonceHeader = "X-Jona-Nonce";
    public const string ResponseSignatureHeader = "X-Jona-Response-Signature";

    public static string PairingCode(ReadOnlySpan<byte> secret) => $"jhcp1_{Base64Url(secret)}";

    public static byte[] ParsePairingCode(string code)
    {
        if (!code.StartsWith("jhcp1_", StringComparison.Ordinal) || code.Length != 49)
            throw new FormatException("Invalid Companion pairing code.");
        var encoded = code[6..].Replace('-', '+').Replace('_', '/');
        var bytes = Convert.FromBase64String(encoded + new string('=', (4 - encoded.Length % 4) % 4));
        if (bytes.Length != 32) throw new FormatException("Invalid Companion pairing code.");
        return bytes;
    }

    public static string SignRequest(ReadOnlySpan<byte> secret, string method, string path, long timestamp, string nonce, string body)
        => Sign(secret, $"{method.ToUpperInvariant()}\n{path}\n{timestamp}\n{nonce}\n{Hash(body)}");

    public static string SignResponse(ReadOnlySpan<byte> secret, int status, string nonce, string body)
        => Sign(secret, $"{status}\n{nonce}\n{Hash(body)}");

    public static bool VerifyRequest(ReadOnlySpan<byte> secret, string method, string path, string body, IHeaderDictionary headers, DateTimeOffset now, out string nonce)
    {
        nonce = headers[NonceHeader].FirstOrDefault() ?? string.Empty;
        var timestampText = headers[TimestampHeader].FirstOrDefault();
        var signature = headers[RequestSignatureHeader].FirstOrDefault();
        if (!long.TryParse(timestampText, out var timestamp) || !IsNonce(nonce) || !IsHexSignature(signature)) return false;
        var current = now.ToUnixTimeSeconds();
        if (timestamp < current - 60 || timestamp > current + 60) return false;
        var expected = SignRequest(secret, method, path, timestamp, nonce, body);
        return CryptographicOperations.FixedTimeEquals(Encoding.ASCII.GetBytes(expected), Encoding.ASCII.GetBytes(signature!.ToLowerInvariant()));
    }

    public static bool VerifyResponse(ReadOnlySpan<byte> secret, int status, string nonce, string body, string? signature)
    {
        if (!IsHexSignature(signature)) return false;
        var expected = SignResponse(secret, status, nonce, body);
        return CryptographicOperations.FixedTimeEquals(Encoding.ASCII.GetBytes(expected), Encoding.ASCII.GetBytes(signature!.ToLowerInvariant()));
    }

    private static string Sign(ReadOnlySpan<byte> secret, string canonical)
        => Convert.ToHexString(HMACSHA256.HashData(secret, Encoding.UTF8.GetBytes(canonical))).ToLowerInvariant();

    private static string Hash(string body)
        => Convert.ToHexString(SHA256.HashData(Encoding.UTF8.GetBytes(body))).ToLowerInvariant();

    private static bool IsNonce(string nonce)
        => nonce.Length >= 20 && nonce.Length <= 24 && nonce.All(c => c is >= 'A' and <= 'Z' or >= 'a' and <= 'z' or >= '0' and <= '9' or '-' or '_');

    private static bool IsHexSignature(string? signature)
        => signature?.Length == 64 && signature.All(c => Uri.IsHexDigit(c));

    private static string Base64Url(ReadOnlySpan<byte> bytes)
        => Convert.ToBase64String(bytes).TrimEnd('=').Replace('+', '-').Replace('/', '_');
}

File: companion/protocol.go
package main

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"errors"
	"net/http"
	"strconv"
	"strings"
	"time"
)

const (
	protocolVersion         = "1"
	pairingPrefix           = "jhcp1_"
	requestSignatureHeader  = "X-Jona-Signature"
	timestampHeader         = "X-Jona-Timestamp"
	nonceHeader             = "X-Jona-Nonce"
	responseSignatureHeader = "X-Jona-Response-Signature"
)

var errInvalidPairingCode = errors.New("invalid Companion pairing code")

func pairingCode(secret []byte) string {
	return pairingPrefix + base64.RawURLEncoding.EncodeToString(secret)
}

func parsePairingCode(code string) ([]byte, error) {
	if !strings.HasPrefix(code, pairingPrefix) || len(code) != len(pairingPrefix)+43 {
		return nil, errInvalidPairingCode
	}
	secret, err := base64.RawURLEncoding.DecodeString(code[len(pairingPrefix):])
	if err != nil || len(secret) != 32 {
		return nil, errInvalidPairingCode
	}
	return secret, nil
}

func signRequest(secret []byte, method, path string, timestamp int64, nonce, body string) string {
	canonical := strings.ToUpper(method) + "\n" + path + "\n" + formatInt(timestamp) + "\n" + nonce + "\n" + bodyHash(body)
	return sign(secret, canonical)
}

func signResponse(secret []byte, status int, nonce, body string) string {
	return sign(secret, formatInt(int64(status))+"\n"+nonce+"\n"+bodyHash(body))
}

func verifyRequest(secret []byte, method, path, body string, headers http.Header, now time.Time) (string, bool) {
	nonce := headers.Get(nonceHeader)
	timestampText := headers.Get(timestampHeader)
	signature := headers.Get(requestSignatureHeader)
	if !validNonce(nonce) || !validSignature(signature) {
		return nonce, false
	}
	timestamp, err := parseInt(timestampText)
	if err != nil {
		return nonce, false
	}
	current := now.Unix()
	if timestamp < current-60 || timestamp > current+60 {
		return nonce, false
	}
	expected := signRequest(secret, method, path, timestamp, nonce, body)
	return nonce, hmac.Equal([]byte(strings.ToLower(signature)), []byte(expected))
}

func verifyResponse(secret []byte, status int, nonce, body, signature string) bool {
	if !validSignature(signature) {
		return false
	}
	return hmac.Equal([]byte(strings.ToLower(signature)), []byte(signResponse(secret, status, nonce, body)))
}

func sign(secret []byte, canonical string) string {
	mac := hmac.New(sha256.New, secret)
	_, _ = mac.Write([]byte(canonical))
	return hex.EncodeToString(mac.Sum(nil))
}

func bodyHash(body string) string {
	hash := sha256.Sum256([]byte(body))
	return hex.EncodeToString(hash[:])
}

func validNonce(nonce string) bool {
	if len(nonce) < 20 || len(nonce) > 24 {
		return false
	}
	for _, char := range nonce {
		if (char < 'A' || char > 'Z') && (char < 'a' || char > 'z') && (char < '0' || char > '9') && char != '-' && char != '_' {
			return false
		}
	}
	return true
}

func validSignature(signature string) bool {
	if len(signature) != 64 {
		return false
	}
	_, err := hex.DecodeString(signature)
	return err == nil
}

func formatInt(value int64) string {
	if value == 0 {
		return "0"
	}
	negative := value < 0
	if negative {
		value = -value
	}
	var digits [20]byte
	index := len(digits)
	for value > 0 {
		index--
		digits[index] = byte('0' + value%10)
		value /= 10
	}
	if negative {
		index--
		digits[index] = '-'
	}
	return string(digits[index:])
}

func parseInt(value string) (int64, error) {
	return strconv.ParseInt(value, 10, 64)
}
